Data review / MongoDB / Docker

jfavlam opened the pull request:
## Quickstart local setup

- `git clone git@github.com:MediaComem/open-co2.git`
- `cd open-co2/server`
- `docker-compose up -d`


Default GQL endpoint is available at http://localhost:4000/graphql

Query example using Apollo Studio (getCategoryByName):
```
query getCategoryByName($name: String!) {
  getCategoryByName(name: $name) {
    title
    fullPath
  }
}
```
Variable:
```
{
  "name": "freight_train"
}
```

---

## Main features/fix

- Setup XLS source file (https://docs.google.com/spreadsheets/d/168Gjlb9d5b_0Vb4q1IeKWGPGBIEcD6oo0ggLZafygwY)
- Seeder to process XLS file
- Persist data in DB (MongoDB)
- Review GraphQL types definition
- Start resolver refactor to get data from DB (all units, all categories, category by name)
- Review project structure (/server/app + /server/seeder)
- Dockerizing project
  - Seed data
  - Populate MongoDB
- GraphQL Playground service in stack (as Apollo alternative) - At http://localhost:4200 by default
- Open CO2 server landing page – At http://localhost:4000 by default

Notes/Missing:
- Resolver and available API queries are incomplete at this stage
- co2eq model will be review to calculate confidence index in mean values + categories with different children units will be ignored

Commit:
chore[INIT]: Add config files

.editorconfig + .prettierrc + nodemon.json

Commit:
feat: Add winston logger

Commit:
build: Add mongodb service

Commit:
feat: Add mongoose models

Commit:
docs: Update .env documentation

Commit:
feat[DATA]: Clean and refactor data that will be used in seeder

Commit:
feat[DATA] Add CO2 data from old file

Temp comment to have less in list

Commit:
docs: Update README

Commit:
feat[UT]: Add Jest to start unit testing project

Commit:
fix: Remove comma in format string function

Commit:
feat: Add function to generate hash from string

Commit:
feat[DATA]: Add input XLS file

Commit:
feat[SEEDER]: Add fileReader

Commit:
feat[SEEDER]: Add fileExporter

Commit:
feat[DB]: Add DB connector using mongoose

Commit:
fix: Update unit model

Commit:
chore: Add default values for env

Commit:
refactor: Remove mocks

Commit:
feat[GRAPHQL]: Update resolver

Commit:
refactor[GRAPHQL]: Update types definition + move scalars

Commit:
feat: Connect to DB when starting server

Commit:
feat: Add temporary content for base express route

Commit:
fix: Check string in formatString util function

Commit:
docs: Add tech stack

Commit:
feat[CLIENT]: Update hoppscotch queries

Commit:
feat[CLIENT]: Add Python client example

Commit:
refactor: Use method shorthand in resolver

Commit:
build[INIT]: Update editor config and add lint script

ESlint + Prettier + Editorconfig

Commit:
fix[MODEL]: Update category model to return an array for CO2eq

CO2eqs can return different CO2eq values with different units

Commit:
feat[RESOLVER]: Resolve children categories

Commit:
feat: Add landing page to access server endpoints

Commit:
fix[DATA]: Update data source

Fix mix CH unit

Commit:
fix[SEEDER]: Clean reader and exporter modules

Commit:
feat[SEEDER]: Add file composer to process the sheets

Need refactor on some methods

Commit:
feat[SEEDER]: Use modules to process XLS file categories

Commit:
refactor[SEEDER]: Refactor and add inline doc to fileReader

Commit:
refactor[SEEDER]: Refactor and add inline doc to fileComposer

Commit:
feat[DATA]: Update XLS with sources

Commit:
feat: Add getDeviationFromArray function

Commit:
docs[METHOD]: Init README for method documentation

Commit:
chore: Update mongo URI

Commit:
docs: Add GitHub cover and update README

Commit:
chore: Split data in separate input/output directories

Commit:
refactor(SEEDER): Add dedicated packages to seeder

Commit:
docs(seeder): Add inline doc to file exporter

Commit:
refactor(seeder): Add dedicated config and utils to seeder

Commit:
refactor(seeder): Update seeder entrypoint

Commit:
refactor(server): Update project app path with dedicated packages

Commit:
refactor: Move .env in dedicated directories app / seeder

Commit:
build: Update Dockerfile with two stages

Seeder -> Server

Commit:
refactor(server): Move all server sources to /app

Commit:
fix(seeder): Update js file path

Commit:
feat(seeder): Add init db script to populate categories

Commit:
build(seeder): Update services to populate DB

Commit:
refactor(server): Move all server source

Commit:
fix: Ignore generated JS file from seeder

Commit:
refactor(seeder)!: Process methods are public

BREAKING CHANGE: fileComposer manage categories OR units sheets

Commit:
fix: Update link to tabular data

Commit:
docs: Move old changelog to server (/app)

Commit:
chore: Change playground default port to 4200

Commit:
fix: Add G_CO2_KWH enum unit type to match sheet

Commit:
docs: Update setup instructions with docker-compose

Commit:
fix(data): Review unit types and descriptions

Commit:
fix: Unit type can be nullable

Commit:
chore(release): 0.1.0

slecorne commented on server/seeder/modules/fileComposer/fileComposer.js:
```diff
@@ -0,0 +1,375 @@
+          const currentLevel = Object.keys(row)[0].replace("Level ", "");
+          row.depth = parseInt(currentLevel);
+          // Delete unused level key
+          delete row[`Level ${n + 1}`];
```
Not sure but I would say the level is n+1
```suggestion
        const currentLevel = n+1;
        const key = `Level ${currentLevel }`;
        if (row[key] !== undefined) {
          // Add title key with value from level
          row.title = row[key];
          // Add depth key from level number
          row.depth = parseInt(currentLevel);
          // Delete unused level key
          delete row[key];
          break;
```

jfavlam replied:
row[key] https://usbeketrica.com/media/83950/download/rocky2.jpg?v=1&inline=1

slecorne commented on server/seeder/modules/fileComposer/fileComposer.js:
```diff
@@ -0,0 +1,375 @@
+   */
+  processCategories() {
+    // 1 – Update object keys from sheet headers
+    this.#updateKeys();
```
```suggestion
    this.#parseRows();
```

slecorne commented on server/seeder/modules/fileComposer/fileComposer.js:
```diff
@@ -0,0 +1,375 @@
+  }
+
+  #addChildrens() {
+    const deepTree = this.#getDeepTree();
```
avoid parsing the tree multiple times. Do once after ParseRows and store in private var

jfavlam replied:
Very good point, especially with this function 👍

slecorne commented on server/seeder/modules/fileComposer/fileComposer.js:
```diff
@@ -0,0 +1,375 @@
+export default class FileComposer {
+  /**
+   * FileComposer constructor
+   * @param {Sheet content from fileReader} sheet
+   */
+  constructor(sheet) {
+    this.sheet = sheet;
```
```suggestion
   * @param {Sheet content from fileReader} rows
   */
  constructor(rows) {
    this.rows= rows;
```

slecorne commented on server/seeder/modules/fileReader/fileReader.js:
```diff
@@ -0,0 +1,58 @@
+/**
+ * Class to read XLS file and extract sheets content
+ */
+export default class FileReader {
```
```suggestion
export default class XLSDataReader{
```

slecorne commented on server/seeder/modules/fileComposer/fileComposer.js:
```diff
@@ -0,0 +1,375 @@
+/**
+ * Class to process XLS sheets and turn it to JS objets
+ */
+export default class FileComposer {
```
```suggestion
export default class DataParser{
```

slecorne commented on server/seeder/modules/fileComposer/fileComposer.js:
```diff
@@ -0,0 +1,375 @@
+    this.#orderObject();
+  }
+
+  /**
+   * Construct and return a deep tree with children categories
+   */
+  #getDeepTree() {
+    const deepTree = [],
+      levels = [deepTree];
+
+    this.sheet.forEach((row) => {
+      levels[row.depth - 1].push({
+        ...row,
+        children: (levels[row.depth] = [])
+      });
+    });
+    return deepTree[0];
+  }
+
+  /**
+   * Find a node in tree by passing its uniqueName
+   * Return the node object (result) and its path (path)
+   */
+  #findInTree(name, tree) {
+    if (tree.uniqueName === name) {
+      const path = [tree.name];
+      return { result: tree, path };
+    } else if (tree.children) {
+      for (const child of tree.children) {
+        const tmp = this.#findInTree(name, child);
+        if (!(Object.keys(tmp).length === 0 && tmp.constructor === Object)) {
+          tmp.path.unshift(tree.name);
+          return tmp;
+        }
+      }
+      return {};
+    }
+  }
```
really hard to follow, I will check if I can help to make it more readable

slecorne reviewed (changes requested):
I put quite a few comments on naming, we can discuss if it really make sense or not

slecorne committed:
update tests

Commit:
fix: Add default NODE_ENV value

Fix issue https://github.com/MediaComem/open-co2/issues/5

Commit:
fix: Remove special characters and keep only letters/numbers to formatString

+ throw Error if no string or string type as input

Commit:
test: Add UT for formatString()

Commit:
docs: Add <img> alt

Commit:
chore: Add sofa api package + graphql tools

Commit:
chore: Rename graphql env and add rest base path

Commit:
refactor: Remove unused utils functions + Dead code in types

Commit:
fix(seeder): Fix typo with row details property name

Commit:
refactor: Rename __dirname conventional CJS name

Commit:
feat(graphql): Use graphql tools to generate executable schema

Commit:
feat(server): Use Sofa and Express router to create REST API

Default base path to /rest

Commit:
refactor(server): Update default env vars

Commit:
refactor(resolver)!: Update queries name for categories/units

BREAKING CHANGE: use categories() and units() queries to get full list of categories and units

Commit:
docs(schema): Add inline documentation for GraphQL schema

Commit:
fix: Update resolver queries name according to Schema

Commit:
fix: Update links on server landing page

Commit:
refactor(server): Manage all vars instantiation in main server function

Commit:
docs: Add section about "Project structure"

Commit:
fix: Update default link to rest endpoint to /infos

Commit:
refactor(env): Use default env vars in a declarative way to avoid declaring "|| default" everywhere in the app

Commit:
chore(init): Move nodemon config to /app + /seeder

Commit:
feat(rest-api)!: Use env var to define tree depth rendered by the rest api

Commit:
refactor(schema): Using query "category" to simplify rest path

REST example to get a category by name http://localhost:4000/rest/category?name=heat

Commit:
refactor(server): Split main init function with dedicated route for landing page

jfavlam committed:
Update server/seeder/modules/fileComposer/fileComposer.js

Co-authored-by: Stéphane Lecorney <stephane.lecorney@heig-vd.ch>

jfavlam committed:
Update server/seeder/modules/fileComposer/fileComposer.js

Co-authored-by: Stéphane Lecorney <stephane.lecorney@heig-vd.ch>

jfavlam committed:
Update server/seeder/modules/fileReader/fileReader.js

Co-authored-by: Stéphane Lecorney <stephane.lecorney@heig-vd.ch>

Commit:
refactor(seeder): Apply naming changes suggested in PR review

Commit:
refactor(seeder): Store deep tree in private var to avoid parsing the tree multiple times

Commit:
Merge branch 'develop' into tests

# Conflicts:
#	server/seeder/modules/DataParser/DataParser.spec.js
#	server/seeder/modules/DataParser/DataParserConsumer.js

Commit:
feat(REST API): Add swagger-ui package to serve REST API doc

Commit:
feat(REST API): Use Sofa/OpenAPI to generate REST doc + server with swagger UI

Commit:
fix(seeder): Use getMeanFromArray() in deviation calculation

Commit:
test: Add utils function test (+ inline doc)

Commit:
test(seeder): Remove unused consumers class

Commit:
test(seeder): Tests for XLSDataReader class

Commit:
test(seeder): Test dataParser.processUnits() method

Commit:
refactor(seeder): Call XLSX read method once in constructor and set private workbook

Commit:
refactor: Review data parser class

Commit:
test: Dummy test for file exporter

Commit:
Merge branch 'develop' into feature/rest-api

jfavlam committed:
Merge pull request #7 from MediaComem/feature/rest-api

REST API

Commit:
ci(tests): Add config to run test on push/PR on develop

Commit:
docs: Update inline doc for file exporter class

Commit:
fix(ci): Change to npm i

Commit:
fix(ci): Update working directory

Commit:
fix(ci): Remove change directory

jfavlam committed:
Merge pull request #6 from MediaComem/tests

update tests
